### DISCUSSION TITLE
Typescript?

### DISCUSSION THREAD

**Discussion by PabloCorso** [elmadev/elmaonline-site#426] [General]
Any thoughts on adding typescript to the project? Could it be a good thing?

Add typescript to a create-react-app project:
https://create-react-app.dev/docs/adding-typescript/

The setups seems very simple, not sure if there would be unexpected complications after it. 

The big disadvantage that I can think of (assuming no big technical difficulties or drawbacks) would be that some people might not have any experience with typescript, and would definitely complex the development for them. Although we can always help those people and everyone wins :>

The advantages are huge (read on https://www.typescriptlang.org/). A key thing would be the possibility of Gradual **Adoption**:
> TypeScript’s type inference means that you don’t have to annotate your code until you want more safety.

🐻

**Comment by ville-j** [comment]
I think it adds a bit of overhead and makes development slower, I'm not sure about the benefits for this project. But of course if the trade is alright it could result in better and safer code.

**Comment by PabloCorso** [comment]
The benefits of type safety done right are extremely profitable. Less errors and much more confident changes regarding types. No half way or outdated PropTypes definitions. Also most popular libraries have types definitions, so the development experiences gets nicer overall.

Could see in this case of JS vs adopting TS as keep doing JS to code "quick-and-dirty" or gradually adopt TS to see great benefits on the long run. Obviously quick-and-dirty is an oxymoron when talking about code, unless you mean to discard the code or not reuse it after making it. Meaning that code written quickly and easily can prove to slow you down a lot in the short and long term. But I'm kind of exaggerating a bit since JS is not a that bad :D

Anyway, I agree that adopting TS would make overhead, and if it's the case that no other active developers know TS or are not very into learning it to keep collaborating to the project, maybe not worth it. 

In my experience, the first time I tried making a side TS project alone by myself, I got really frustrated and could not finish it. I go stuck and it sucked. Currently I got to understand it mostly and I love it. 

Having said that, we are not alone here. So, in case we could at least consider TS for this project, there could be ways to mitigate the overhead and frustration for others:
- Gradual Adoption. For example, introducing it in new areas or the ones that could need it the most. And although we should not use the `any` type to implicitly define types, importing JS files without types into TS files or vice versa should not conflict. In some way, trying to ease the adoption as much as possible, although someone will experience overhead at some point for sure :s
- Create good examples inside the code for others. Rewrite some files into TS with good examples for components, functions, constants, etc. Where to define types? Use interfaces or types? 
- Help in Discord :> and make good tips like always try letting typescript infer types unless you need to define it yourself and how to easily know what type is what 🐻

**Comment by PabloCorso** [comment]
Did a little test in elmaonline-web project, converting a single file to TS `src\pages\battle\index.tsx` and the setup was straightforward. But got the following errors from imports:
```ts
import Layout from 'components/Layout';
import { battleStatus } from 'utils/battle';

Cannot find module 'components/Layout' or its corresponding type declarations.ts(2307)
```

Tried updating the `tsconfig.json` with `"paths"` property to match those aliases in the overwritten webpack config, and got the following result:

```
$ react-app-rewired start
The following changes are being made to your tsconfig.json file:
  - compilerOptions.paths must not be set (aliased imports are not supported)
```

So for that project I thought of possible fixes for this:
- Don't use aliases inside TS files.
- Find another way to make it work dunno :D

**Comment by j-masci** [comment]
i'm not against typescript but don't have a strong opinion either way at this point.

**Comment by sunehs** [comment]
I don't really see any major benefits in typescript personally. There's strong typing which is fine but that's basically it which then adds complexity and an extra transpile layer.

Other benefits that are usually mentioned is basically extra linting which should be possible without typescript.

**Comment by PabloCorso** [comment]
Well in that case if people are not interested in it, and the fact that not everyone will want to invest on it, then it won't happen.

But for me that's the only downside for using typescript in this project. Type safety it's not just fine :/ it is great. It gives so much better 
and faster development experience, with less trivial typos and mistakes, and so much more confidence to change and use other's code. Consider doing a big (but necessary) refactor with and without type check, there is a big difference.

But again, the development team skills or language knowledge and preferences are always a strong argument to consider. Same as for including any dependency in a project. So that's that.

**Comment by sunehs** [comment]
Having worked on a typescript project this year, I've now seen the light. It's probably unlikely we can get to full strict mode any time soon, but getting the autocomplete from libraries and converting the most used components/functions would be a nice help in future work.

For current contributers: are you familiar with typescript and do you have any thoughts on this?

**Comment by abulaatikko** [comment]
I haven't seen the light yet but my employer uses TS - I haven't really coded much frontend recently but I think consensus is pro TypeScript among frontend developers. Maybe it's good in healthy code base and when done correct? To my eye it decreases the code readability. Simplicity is a virtue in programming. GPT recommends in large codebases with several developers.
